homeassistant/components/saj/sensor.py

```diff
@@ -0,0 +1,161 @@
+"""SAJ solar inverter interface."""
+import asyncio
+from datetime import date, timedelta
+import logging
+
+import pysaj
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_HOST, CONF_SCAN_INTERVAL
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+SENSOR_ICON = "mdi:solar-power"
+
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend(
+        {vol.Required(CONF_HOST): cv.string}, extra=vol.PREVENT_EXTRA
+    )
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up SAJ sensors."""
+
+    # Init all sensors
+    sensor_def = pysaj.Sensors()
+
+    # Use all sensors by default
+    hass_sensors = []
+    used_sensors = []
+
+    for sensor in sensor_def:
+        hass_sensors.append(SAJsensor(sensor))
+        used_sensors.append(sensor)
+
+    saj = pysaj.SAJ(config[CONF_HOST])
+
+    async_add_entities(hass_sensors)
+
+    backoff = 0
+    backoff_step = 0
+
+    async def async_saj(event):
+        """Update all the SAJ sensors."""
+        nonlocal backoff, backoff_step
+        tasks = []
+        # If reading sensors has failed multiple times
+        if backoff > 1:
+            backoff -= 1
+
+            # SAJ inverters are powered by DC via solar panels and thus are
+            # offline after the sun has set. If a sensor resets on a daily
+            # basis like "today_yield", this reset won't happen automatically.
+            # Code below checks if today > day when sensor was last updated
+            # and if so: set state to STATE_UNKNOWN.
+            # Sensors with live values like "temperature" or "current_power"
+            # will also be reset to STATE_UNKNOWN.
+            for sensor in hass_sensors:
+                if (sensor.per_day_basis and date.today() > sensor.date_updated) or (
+                    not sensor.per_day_basis and not sensor.per_total_basis
+                ):
+                    task = sensor.async_update_values(unkown_state=True)
+                    if task:
+                        tasks.append(task)
+            if tasks:
+                await asyncio.wait(tasks)
+
+            return
+
+        values = await saj.read(used_sensors)
+        if not values:
+            try:
+                backoff = [1, 1, 1, 6, 30][backoff_step]
```

Comment:
What is the idea with the backoff list? How does this relate to the time between tries? Can we rewrite this to use a change in interval and reschedule the next update call instead. It's hard to follow the current logic.

Reply:
Actually I copied it from the SMA component (first asked @kellerza if it was ok to base my component on his code).
I thought it was a good idea to not constantly try and update the sensor values if it fails and increase the the interval each time it fails for a certain time and then start over.
But if you say it's possible to change the update interval at runtime then this would even be better.
What would be the best approach to change the update interval dynamically?

Reply:
Instead of using `async_track_time_interval` to update at a constant interval, use `async_call_later` to just schedule the next update call after interval. Save the return value of the `async_call_later` call, to be able to cancel the next scheduled update. Then we can dynamically control the time of the next update as needed, and there's a clear connection between the backoff mechanism and the time of the next update.

Reply:
Actually I want to update SMA to use the backoff mechanism I wrote for DuckDNS

Have a look at DuckDNS mechanism and according to me this is a clear separation of the work and the backoff state, even handling thrown exceptions by the work function.
I am not a big fan of call_later, since you still have to keep your own state in each implementation

Reply:
Modifying the position in the list of interval times instead of modifying the interval time itself in the backoff mechanism makes the logic more complex.

My advice is to think about how specific we need to be when selecting the interval times and weigh that against the increase in code readability.

Reply:
After some help I've got the DuckDNS backoff implementation working (thanks again @kellerza).
I hope this meets the requirements that you've asked?

Reply:
It's ok. I would skip the list of interval times and modify the time by multiplying it on each failure, capping it with `min` and resetting it on success.
Something like this:
```py
interval = min(interval * 2, max_time)
```

This is usually good enough. But it's ok with the list of times if you prefer that.

Reply:
Good proposition, I've implemented the change as you suggested!